### DISCUSSION TITLE
[Single-phase pipe flow] Dirackernel oscillation?

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#24011] [Q&A General] [answered]
Dear all,

Recently, I wanted to simulate a 1D single-phase pipe flow with pressure (Pa), temperature (K), and volume flow rate (m3/s) as variables. I added a Dirackernel of pressure to remove the fluid mass at the middle position and ending position, respectively.  However, the results are unstable.  

Could you please help me on fixing this kind of Dirackernel oscillation? Thanks a lot.

Regards,

J

**Results of the dirackernel of pressure at the middle position**

![image](https://user-images.githubusercontent.com/56687392/230796608-803e6cf2-729e-4f0e-9183-66c75c4ce603.png)

**Results of the dirackernel of pressure at the ending position**

![image](https://user-images.githubusercontent.com/56687392/230796636-d4ce3ecd-8941-495b-a059-8982e9243051.png)

**Comment by GiudGiud** [comment]
Hello

Why a DiracKernel and not just a boundary condition at the end, or a volumetric sink in the middle?

Guillaume

**Reply by Joseph-0123** [reply]
> Hello
> 
> Why a DiracKernel and not just a boundary condition at the end, or a volumetric sink in the middle?
> 
> Guillaume

Thanks for your answers. 

What is the difference between DiracKernel and a volumetric sink in the middle? I know the DiracKernel plays as a point source. How does the role volumetric sink play?

**Reply by GiudGiud** [reply]
a volumetric source (or sink) would be smoother numerically than a DiracKernel (adding only on one node). Less discontinuities tend to bring less oscillations

**Reply by Joseph-0123** [reply]
> a volumetric source (or sink) would be smoother numerically than a DiracKernel (adding only on one node). Less discontinuities tend to bring less oscillations

So you mean the volumetric source (or sink) is adding to several nodes rather than one node? How many nodes does it apply to? How can I control the number of nodes?

**Reply by GiudGiud** [reply]
well it really would apply to elements rather than nodes. 
Are you saying nodes because that equation is using nodal vartiables?

**Reply by Joseph-0123** [reply]
> Are you saying nodes because that equation is using nodal vartiables?

Yes, because FE equations use nodes to compute variables. I am wondering how does this volumetric sink work on the element in the FE equations or weak forms?

**Reply by Joseph-0123** [reply]
> well it really would apply to elements rather than nodes. Are you saying nodes because that equation is using nodal vartiables?

Sorry, I cannot find the introduction page for using a volumetric source (or sink) in MOOSE. Could you please share the links and examples? Thanks.

**Reply by GiudGiud** [reply]
> Yes, because FE equations use nodes to compute variables.

Typically not, variables that live on nodes are evaluated on quadrature points within each element, to compute the contribution to the residual/jacobian of each kernel. 
You could just try it? BodyForce is an example of a volumetric source term kernel

**Reply by Joseph-0123** [reply]
> volumetric

So the volumetric source (or sink) plays the role on the element through the quadrature points, right?

**Reply by Joseph-0123** [reply]
> > Yes, because FE equations use nodes to compute variables.
> 
> Typically not, variables that live on nodes are evaluated on quadrature points within each element, to compute the contribution to the residual/jacobian of each kernel. You could just try it? BodyForce is an example of a volumetric source term kernel



> > Yes, because FE equations use nodes to compute variables.
> 
> Typically not, variables that live on nodes are evaluated on quadrature points within each element, to compute the contribution to the residual/jacobian of each kernel. You could just try it? BodyForce is an example of a volumetric source term kernel

Dear Guillaume,

I follow your suggestion and replace dirackernel at the middle point by adding the volumetric sink (i.e., BodyForce kernel) for the middle element (i.e., the 10m part 'block 1' on the middle position of the pipe).

  [./sink] 
    type = BodyForce
    variable = p
    function = -600
    block = 1
  [../]

However, I also get unstable results (shown as follows) as the dirackernel. By the way, I also set pressure BCs (i.e., NeumannBC) to remove the fluid mass in the middle position of the pipe. I still get similar unstable results.

  [./sink]
    type = FunctionNeumannBC
    variable = p
    boundary = out_flow
    function = -1800
  [../]

Could you please help me on fixing this oscillation? Thanks a lot.

Regards,

J

**one element as 'block 1' with a 10m length**

![95ec2598bf3ffae5dcbb1f1d063b946](https://user-images.githubusercontent.com/56687392/230956608-c7cf491a-15b8-4723-adb8-25a2ce356e37.png)

**2 times refined element as 'block 1' with a 10m length**

![6056147a5c7f2a67650b507c3fbc413](https://user-images.githubusercontent.com/56687392/230957427-d8693fec-9ebe-49dd-b5a5-1e478e190108.png)

**Reply by GiudGiud** [reply]
ok seems there are other issues. 
What does your variables block look like? Are you using THM or the navier stokes module?

**Reply by Joseph-0123** [reply]
> ok seems there are other issues. What does your variables block look like? Are you using THM or the navier stokes module?

Here are my variables block. I am using our in-house-developed [app](https://github.com/MGK-Lab/moskito) somewhat similar to the Navier stokes module.

[Variables]
  [./volumetric_flow_rate]
    scaling = 1e-10
  [../]
  [./pressure]
    scaling = 1e-2
  [../]
  [./Temperature]
    scaling = 1e-14
  [../]
[]

**Reply by GiudGiud** [reply]
first order for both pressure and flow rate / velocity tends to require stabilization.
You could try second order for velocities?

**Reply by Joseph-0123** [reply]
> first order for both pressure and flow rate / velocity tends to require stabilization. You could try second order for velocities?

For the 1D MOOSE mesh, it reminds me that "ERROR: Bad ElemType = EDGE2 for SECOND order approximation!
". So I use Gmsh to create the 1D mesh with 2nd order. I set the 2nd order for volumetric_flow_rate as shown below. The results get steady but wrong. If I set the 2nd order for volumetric_flow_rate and pressure, the results are unstable and wrong.
```
[Variables]
  [./volumetric_flow_rate]
    scaling = 1e-10
    order = SECOND
    family = LAGRANGE
  [../]
  [./pressure]
    scaling = 1e-2
  [../]
  [./Temperature]
    scaling = 1e-14
  [../]
[]
```
**2nd order for volumetric_flow_rate**

![00d22ef906acc23c56e2a8f5fd43f5f](https://user-images.githubusercontent.com/56687392/230999697-690b9d72-5b03-40ce-8d14-62ca1b38d4b1.png)

**2nd order for volumetric_flow_rate and pressure**

![1681161213433](https://user-images.githubusercontent.com/56687392/231000061-75f21fcd-4e3d-41aa-ae6e-8f87c5d8922d.png)

**Reply by GiudGiud** [reply]
>  If I set the 2nd order for volumetric_flow_rate and pressure

Ok this is expected, you are back to same order for velocity and pressure
I think where you want to be is 1 order higher for flow rate.

**Reply by Joseph-0123** [reply]
> > If I set the 2nd order for volumetric_flow_rate and pressure
> 
> Ok this is expected, you are back to same order for velocity and pressure I think where you want to be is 1 order higher for flow rate.

If I set 1 order higher for the flow rate and temperature than Pressure, the results are somehow more stable than the initial results but wrong.

```
[Variables]
[./volumetric_flow_rate]
scaling = 1e-10
[../]
[./pressure]
scaling = 1e-2
order = SECOND
family = LAGRANGE
[../]
[./Temperature]
scaling = 1e-14
[../]
[]
```
![1681198156623](https://user-images.githubusercontent.com/56687392/231088279-e05f4014-985f-4522-a3c8-629dc70add81.png)

**Reply by GiudGiud** [reply]
> If I set 1 order higher for the flow rate and temperature than Pressure, the results are somehow more stable than the initial results but wrong.

what you are doing here is the opposite, one order higher for pressure

**Reply by Joseph-0123** [reply]
> > If I set 1 order higher for the flow rate and temperature than Pressure, the results are somehow more stable than the initial results but wrong.
> 
> what you are doing here is the opposite, one order higher for pressure

Ok. I set the 2nd order for volumetric_flow_rate as shown below. The results get steady but wrong. 
```
[Variables]
[./volumetric_flow_rate]
scaling = 1e-10
order = SECOND
family = LAGRANGE
[../]
[./pressure]
scaling = 1e-2
[../]
[./Temperature]
scaling = 1e-14
[../]
[]
```
**2nd order for volumetric_flow_rate**

![image](https://user-images.githubusercontent.com/56687392/231175962-005c7e70-714b-472a-84cb-7e8a1f83c989.png)

**Reply by GiudGiud** [reply]
The boundary conditions look wrong. 
What else is wrong here?

**Reply by Joseph-0123** [reply]
> The boundary conditions look wrong. What else is wrong here?

The BCs and ICs are shown here. The pressure, flow rate, and T are wrong due to their negative values.

[BCs]
  [./top_q]
    type = FunctionDirichletBC
    variable = q
    boundary = right
    function = flowrate_ini
  [../]
  [./bottom_p]
    type = FunctionDirichletBC
    variable = p
    boundary = left
    function = hydrostatic
  [../]
  [./top_T]
    type = FunctionDirichletBC
    variable = T
    boundary = right
    function = temperature_ini
  [../]
[]

[ICs]
  [./flowrate_ic]
    type = FunctionIC
    variable = q
    function = flowrate_ini
  [../]
  [./pressure_ic]
    type = FunctionIC
    variable = p
    function = hydrostatic
  [../]
  [./temperature_ic]
    type = FunctionIC
    variable = T
    function = temperature_ini
  [../]
[]

[Functions]
  [./flowrate_ini]
    type = ParsedFunction
    expression = 4e-2
  [../]
  [./hydrostatic]
    type = ParsedFunction
    expression = 1e6-9810*x
  [../]
  [./temperature_ini]
    type = ParsedFunction
    expression = 293.15-0.03*x
  [../]
[]

**Reply by GiudGiud** [reply]
Yeah something s wrong. 
What fluid properties are you using? Can you try constant ones?

I would go back to a single component / single section system, get that one right. Then you can try this problem with 5 sections

**Reply by Joseph-0123** [reply]
> Yeah something s wrong. What fluid properties are you using? Can you try constant ones?
> 
> I would go back to a single component / single section system, get that one right. Then you can try this problem with 5 sections

1. I use the fluid property of Brine. When I try constant ones, the results are somehow more stable than the initial results but wrong.


![1681235474828](https://user-images.githubusercontent.com/56687392/231247748-9dbe2cce-9574-4be5-bbeb-122f61ebdc4a.png)

**Reply by Joseph-0123** [reply]
> Yeah something s wrong. What fluid properties are you using? Can you try constant ones?
> 
> I would go back to a single component / single section system, get that one right. Then you can try this problem with 5 sections

2.Yes, I simulate a single-component system. Initially, I only have a section and do not use the dirackernel and Bodyforce kernel. The results shown as below are stable and correct. Once the dirackernel or Bodyforce kernel is added, the stabilization problem and even error results occur.


**Results of single section system with no dirackernel and Bodyforce kernel**
![1681235717293](https://user-images.githubusercontent.com/56687392/231248224-3e651d3a-3346-41b9-9bad-12d250d4d39e.png)

**Reply by GiudGiud** [reply]
@joshuahansel any tips on adding a volume term to 1D flow

**Reply by joshuahansel** [reply]
I don't think there is any MOOSE-related question remaining - kernel and dirac kernel are what you would use for sources. The question is one about stabilization and the particular formulation, and those I do not have the expertise to answer. The only thing I can say is that without proper stabilization, different combinations of source terms and boundary conditions can cause oscillatory behavior such as this.

**Reply by Joseph-0123** [reply]
> I don't think there is any MOOSE-related question remaining - kernel and dirac kernel are what you would use for sources. The question is one about stabilization and the particular formulation, and those I do not have the expertise to answer. The only thing I can say is that without proper stabilization, different combinations of source terms and boundary conditions can cause oscillatory behavior such as this.

Hello Guillaume and Joshua, thanks for your help.

It looks like an effective method for solving this oscillation problem is adding proper stabilization terms in my mass/momentum/energy kernels, right? Could you please recommend some introductions and examples of adding stabilization terms for pipe flow? Thanks a lot.

**Reply by GiudGiud** [reply]
We have some documentation of SUPG stabilization in the NS module. 
Is this where you want to go though?
I would try switching to finite volume at this point, we have not needed stabilization with it and it's proven quite resilient

**Reply by Joseph-0123** [reply]
> We have some documentation of SUPG stabilization in the NS module. Is this where you want to go though? I would try switching to finite volume at this point, we have not needed stabilization with it and it's proven quite resilient

I am sorry that I am not familiar with the finite volume method.  So I do not know the workloads for switching from finite element to finite volume for simulating pipe flow. 

Do you think it is easy and time-saving for these changes compared to adding SUPG stabilization terms in my old mass/momentum/energy kernels?

**Reply by GiudGiud** [reply]
I dont know if it's time saving.
A few other options:
- use the thermal hydraulics module in MOOSE that has 1D FE with RDG
- apply for the export controlled SAM code on the INL NCRC website that has 1D FE with discretization

**Reply by Joseph-0123** [reply]
> I dont know if it's time saving. A few other options:
> 
> * use the thermal hydraulics module in MOOSE that has 1D FE with RDG
> * apply for the export controlled SAM code on the INL NCRC website that has 1D FE with discretization

Thanks for your suggestions, Guillaume. I will learn and check these apps.

**Reply by joshuahansel** [reply]
I can speak for the flow model in the thermal hydraulics module (THM). What is it exactly that you're trying to simulate? You were removing mass in the *middle* of the channel? Are you trying to do some kind of pipe break? This is something we haven't tried yet in THM, as far as I'm aware. It might be possible with source terms (Kernels or DiracKernels). If that doesn't work, I'd probably actually break the channel into two components and use a special junction. We don't yet have any special "break" junction, but at the very least, you could use a volume junction with 3 channels: pipe half 1, pipe half 2, and then a dummy pipe for the break, connected to an outlet. Let me know if you end up going that route, and I can help.

**Reply by Joseph-0123** [reply]
> I can speak for the flow model in the thermal hydraulics module (THM). What is it exactly that you're trying to simulate? You were removing mass in the _middle_ of the channel? Are you trying to do some kind of pipe break? This is something we haven't tried yet in THM, as far as I'm aware. It might be possible with source terms (Kernels or DiracKernels). If that doesn't work, I'd probably actually break the channel into two components and use a special junction. We don't yet have any special "break" junction, but at the very least, you could use a volume junction with 3 channels: pipe half 1, pipe half 2, and then a dummy pipe for the break, connected to an outlet. Let me know if you end up going that route, and I can help.


Hello Joshua, thanks for your suggestions.

1. I use our in-house-developed [app](https://github.com/MGK-Lab/moskito) somewhat similar to the Navier stokes module to simulate a 1D single-phase and single-component pipe flow with pressure (Pa), temperature (K), and volume flow rate (m3/s) as variables, not THM. 
 
2. A leakage point/part in the middle of the pipe is assumed to lose fluid mass and energy. But the mechanical causes/processes of this leakage are not included, only the resulting leakage phenomenon is considered.
 
3. Initially, if the leakage is not included, the results of the pipe flow are stable and correct. Once the dirackernel or Bodyforce kernel is added to mimic the leakage, the stabilization problem and even error results occur.

4. The idea of a volume junction with 3 channels is an option for this problem. But larger errors may also be caused by this artificial segmentation and data transfer between the 3 channels. I will try this method.

Best regards,

J

**Reply by Joseph-0123** [reply]
> We have some documentation of SUPG stabilization in the NS module. Is this where you want to go though? I would try switching to finite volume at this point, we have not needed stabilization with it and it's proven quite resilient

Hello @GiudGiud @joshuahansel , 

Thanks for your suggestions. Maybe I need to use finite volume rather than finite element for avoiding the stabilization problem. 
Could you please recommend any MOOSE modules/applications to solve my problem of a 1D single-phase and single-component compressible pipe flow? Thanks a lot.

**Reply by GiudGiud** [reply, marked as answer]
Hello

We did :
> use the thermal hydraulics module in MOOSE that has 1D FE with RDG
> apply for the export controlled SAM code on the INL NCRC website that has 1D FE with discretization

You could try using the NavierStokes module in 1D but you would have to re-create all the logic for the pipes. It's better to  start from THM or SAM

Guillaume
